src/Database/Traits/ArraySource.php

```diff
@@ -0,0 +1,307 @@
+            }
+
+            throw $e;
+        }
+    }
+
+    /**
+     * Determines the best column schema type from a given value
+     *
+     * @param mixed $value
+     * @return string
+     */
+    protected function resolveArrayDatatype($value): string
+    {
+        if (is_int($value)) {
+            return 'integer';
+        }
+
+        if (is_numeric($value)) {
+            return 'float';
+        }
+
+        if (is_string($value)) {
+            return 'string';
+        }
+
+        if (is_object($value) && $value instanceof \DateTimeInterface) {
+            return 'dateTime';
+        }
+
+        return 'string';
+    }
+
+    /**
+     * Determines if the temporary SQLite database for this model's array records will be stored.
+     */
+    protected function canStoreArrayDb(): bool
+    {
+        // A model may add a $cacheArray property which defines if this model will be cached or not
+        if ($this->propertyExists('cacheArray') && ((bool) $this->cacheArray) === false) {
+            return false;
+        }
+
+        $sourceCacheDir = $this->getArrayDbDir();
+
+        if ($sourceCacheDir === false) {
+            return false;
+        }
+
+        if (!File::exists($sourceCacheDir)) {
+            if (!File::makeDirectory($sourceCacheDir, 0777, true)) {
+                return false;
+            }
+        }
+
+        return File::isWritable($sourceCacheDir);
+    }
+
+    /**
+     * Gets the directory where the array databases will be stored.
+     */
+    protected function getArrayDbDir(): string|false
+    {
+        $sourcePath = Config::get('cms.arraySourcePath', storage_path('framework/cache/array-source/'));
+
+        if ($sourcePath === false) {
+            return false;
+        }
+
+        return realpath($sourcePath);
+    }
+
+    /**
+     * Gets the path where the array database will be stored.
+     *
+     * @return string
+     */
+    protected function getArrayDbPath(): string
+    {
+        return $this->getArrayDbDir() . '/' . Str::kebab(static::class) . '.sqlite';
+    }
+
+    /**
+     * Determines if the stored array DB should be updated.
+     *
+     * @return boolean
+     */
+    protected function arrayDbNeedsUpdate(): bool
+    {
+        if (!$this->canStoreArrayDb()) {
+            return true;
+        }
+
+        if (!File::exists($this->getArrayDbPath())) {
+            return true;
+        }
+
+        $modelFile = (new ReflectionClass(static::class))->getFileName();
+
+        if (File::lastModified($this->getArrayDbPath()) < File::lastModified($modelFile)) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Sets the array chunk size when storing inserts.
+     *
+     * Sometimes, SQLite will complain if given too many records to insert at once, so we will split the records up
+     * into reasonable chunks and insert them in groups.
+     *
+     * @return int
+     */
+    protected function getArrayChunkSize()
```

Comment:
Trait method names should be prefixed with the namespacing, so `arraySourceGetChunkSize()`, `arraySourceDbNeedsUpdate()`, etc.

Reply:
to avoid name clashing, yeah, good point.

Reply:
Alright, but just to note, none of our other traits follow that yet.